src/lib/custom-resources/cdk-s3-update-logarchive-bucket-policy/runtime/src/index.ts

```diff
@@ -0,0 +1,201 @@
+const logArchiveReadOnlySid = 'SSM Log Archive Read Only Roles';
+
+interface PolicyStatement {
+  Sid: string;
+  Effect: string;
+  Action: string[];
+  Principal: {
+    AWS: string[];
+  };
+  Resource: string[] | string;
+}
+
+async function onEvent(event: CloudFormationCustomResourceEvent) {
+  console.log(`Adding roles with /'ssm-log-archive-read-only-access: true/'
+   to the Log Archive Bucket Policy...`);
+
+  // eslint-disable-next-line default-case
+  switch (event.RequestType) {
+    case 'Create':
+      return onCreate(event);
+    case 'Update':
+      return onUpdate(event);
+    case 'Delete':
+      return onDelete(event);
+  }
+}
+
+async function getBucketPolicy(logBucketName: string) {
+  try {
+    const response = await s3
+      .getBucketPolicy({
+        Bucket: logBucketName,
+      })
+      .promise();
+    if (response.Policy) {
+      return JSON.parse(response.Policy);
+    }
+    return {};
+  } catch (err) {
+    console.error(err, err.stack);
+    throw err;
+  }
+}
+
+async function putBucketPolicy(logBucketName: string, policy: string) {
+  try {
+    await s3
+      .putBucketPolicy({
+        Bucket: logBucketName,
+        Policy: policy,
+      })
+      .promise();
+  } catch (err) {
+    console.error(err, err.stack);
+    throw err;
+  }
+}
+
+async function getKmsKeyPolicy(keyArn: string | undefined) {
+  if (keyArn) {
+    try {
+      const response = await kms
+        .getKeyPolicy({
+          KeyId: keyArn,
+          PolicyName: 'default',
+        })
+        .promise();
+      if (response.Policy) {
+        return JSON.parse(response.Policy);
+      }
+      return {};
+    } catch (err) {
+      console.error(err, err.stack);
+      throw err;
+    }
+  }
+  return {};
+}
+
+async function putKmsKeyPolicy(keyArn: string | undefined, policy: string) {
+  if (keyArn) {
+    try {
+      await kms
+        .putKeyPolicy({
+          KeyId: keyArn,
+          Policy: policy,
+          PolicyName: 'default',
+        })
+        .promise();
+    } catch (err) {
+      console.error(err, err.stack);
+      throw err;
+    }
+  } else {
+    console.error('No KMS Key configured for this bucket');
+  }
+}
+
+function removeExistingReadOnlyStatement(statements: PolicyStatement[]) {
+  return statements.filter(statement => statement['Sid'] !== logArchiveReadOnlySid);
+}
+
+async function addStatementToPolicy(policy: any, statement: PolicyStatement) {
+  if (Object.keys(policy).length > 0) {
+    const updatedStatements = removeExistingReadOnlyStatement(policy.Statement);
+    updatedStatements.push(statement);
+    policy.Statement = updatedStatements;
+    return policy;
+  } else {
+    return {
+      Version: '2012-10-17',
+      Statement: [statement],
+    };
+  }
+}
+
+async function createOrUpdateBucketPolicy(props: HandlerProperties) {
+  let bucketPolicy = await getBucketPolicy(props.logBucketName);
+  let keyPolicy = await getKmsKeyPolicy(props.logBucketKmsKeyArn);
+
```

Comment:
Add extra check for ` props.roles ` to avoid error if there is no roles defined with ` 'ssm-log-archive-read-only-access': true `.

Reply:
Fixed in latest commit (and I can't believe I forgot to cover that use case!).